server/src/main/java/org/elasticsearch/action/admin/cluster/state/ClusterStateResponse.java

```diff
@@ -55,14 +60,74 @@ public ClusterName getClusterName() {
         return this.clusterName;
     }
 
+    /**
+     * Returns whether the request timed out waiting for a cluster state with a metadata version equal or
+     * higher than the specified metadata.
+     */
+    public boolean isWaitForTimedOut() {
+        return waitForTimedOut;
+    }
+
     public ClusterStateResponse(StreamInput in) throws IOException {
         clusterName = new ClusterName(in);
-        clusterState = ClusterState.readFrom(in, null);
+        clusterState = in.readOptionalWriteable(innerIn -> ClusterState.readFrom(innerIn, null));
+        if (in.getVersion().onOrAfter(Version.V_4_4_0)) {
+            waitForTimedOut = in.readBoolean();
+        }
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         clusterName.writeTo(out);
-        clusterState.writeTo(out);
+        out.writeOptionalWriteable(clusterState);
```

Comment:
Doesn't this change the streaming? `writeOptional` adds a boolean to indicate whether it's null or not - that wasn't present before.

Reply:
Thanks for finding this! `clusterState` can be null here, therefore it was changed. So I guess this should be then checked with the version and the streaming adapted accordingly.